searches/binary_search.py

```diff
@@ -297,12 +282,78 @@ def binary_search_by_recursion(
         return binary_search_by_recursion(sorted_collection, item, midpoint + 1, right)
 
 
+def exponential_search(sorted_collection: list[int], item: int) -> int:
+    """Pure implementation of an exponential search algorithm in Python
+    Resources used:
+    https://en.wikipedia.org/wiki/Exponential_search
+
+    Be careful collection must be ascending sorted otherwise, result will be
+    unpredictable
+
+    :param sorted_collection: some ascending sorted collection with comparable items
+    :param item: item value to search
+    :return: index of the found item or -1 if the item is not found
+
+    the order of this algorithm is O(lg I) where I is index position of item if exist
+
+    Examples:
+    >>> exponential_search([0, 5, 7, 10, 15], 0)
+    0
+    >>> exponential_search([0, 5, 7, 10, 15], 15)
+    4
+    >>> exponential_search([0, 5, 7, 10, 15], 5)
+    1
+    >>> exponential_search([0, 5, 7, 10, 15], 6)
+    -1
+    """
+    if list(sorted_collection) != sorted(sorted_collection):
+        raise ValueError("sorted_collection must be sorted in ascending order")
+    bound = 1
+    while bound < len(sorted_collection) and sorted_collection[bound] < item:
+        bound *= 2
+    left = bound // 2
+    right = min(bound, len(sorted_collection) - 1)
+    last_result = binary_search_by_recursion(
```

Comment:
@tianyizheng02 `exponential_search()` embeds a call to `binary_search_by_recursion()` but as the benchmarks demonstrate delivers a faster result.  Looking at the first commit makes this clearer.